File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.py
import heapq

def solution(n, k, enemy):
    answer = 0
    defence = []
    for i in range(len(enemy)):
        n -= enemy[i]
        heapq.heappush(defence, -enemy[i])
        if n < 0:
            if k <= 0:
                break
            
            k -= 1
            n += -heapq.heappop(defence)
        
        answer += 1
            
    return answer

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int i = 0;
    priority_queue<int> defence;
    
    for(; i < enemy.size(); i++){
        n -= enemy[i];
        defence.push(enemy[i]);
        
        if(n < 0){
            if(k <= 0) break;
            
            k--;
            n += defence.top();
            defence.pop();
        }
    }
    return i;
}